src/inference/InterfaceExtensions.jl

```diff
@@ -0,0 +1,334 @@
+    local names
+    vals  = Vector{Vector{Float64}}()
+    for t in ts
+        names, vs = flatten_namedtuple(t.θ)
+        push!(vals, vs)
+    end
+    
+    return names, vals
+end
+
+function flatten_namedtuple(nt::NamedTuple{pnames}) where {pnames}
+    vals  = Vector{Float64}()
+    names = Vector{AbstractString}()
+    for k in pnames
+        v = nt[k]
+        if length(v) == 1
+            flatten(names, vals, string(k), v)
+        else
+            for (vnval, vn) in zip(v[1], v[2])
+                flatten(names, vals, vn, vnval)
+            end
+        end
+    end
+    return names, vals
+end
+
+function flatten(names, value :: Array{Float64}, k :: String, v)
+    if isa(v, Number)
+        name = k
+        push!(value, v)
+        push!(names, name)
+    elseif isa(v, Array)
+        for i = eachindex(v)
+            if isa(v[i], Number)
+                name = string(ind2sub(size(v), i))
+                name = replace(name, "(" => "[");
+                name = replace(name, ",)" => "]");
+                name = replace(name, ")" => "]");
+                name = k * name
+                isa(v[i], Nothing) && println(v, i, v[i])
+                push!(value, Float64(v[i]))
+                push!(names, name)
+            elseif isa(v[i], Array)
+                name = k * string(ind2sub(size(v), i))
+                flatten(names, value, name, v[i])
+            else
+                error("Unknown var type: typeof($v[i])=$(typeof(v[i]))")
+            end
+        end
+    else
+        error("Unknown var type: typeof($v)=$(typeof(v))")
+    end
+    return
+end
+
+# Default Chains constructor.
+function Chains(
+    rng::AbstractRNG,
+    model::ModelType,
+    spl::Sampler,
+    N::Integer,
+    ts::Vector{T};
+    discard_adapt::Bool=true,
+    save_state=false,
+    kwargs...
+) where {ModelType<:Sampleable, T<:AbstractTransition}
+    # Check if we have adaptation samples.
+    if discard_adapt && :n_adapts in fieldnames(typeof(spl.alg))
+        ts = ts[(spl.alg.n_adapts+1):end]
+    end
+
+    # Convert transitions to array format.
+    # Also retrieve the variable names.
+    nms, vals = _params_to_array(ts, spl)
+
+    # Get the extra field names from the sampler state type.
+    # This handles things like :lp or :weight.
+    extra_params = additional_parameters(T)
+
+    # Get the values of the extra parameters.
+    extra_values = vcat(map(t -> [getproperty(t, p) for p in extra_params], ts))
+
+    # Extract names & construct param array.
+    nms = string.(vcat(nms..., string.(extra_params)...))
+    parray = vcat([hcat(vals[i]..., extra_values[i]...) for i in 1:length(ts)]...)
+
+    # If the state field has average_logevidence or final_logevidence, grab that.
+    le = missing
+    if :average_logevidence in fieldnames(typeof(spl.state))
+        le = getproperty(spl.state, :average_logevidence)
+    elseif :final_logevidence in fieldnames(typeof(spl.state))
```

Comment:
What is `final_logevidence`? Is this needed?

Reply:
Yes, it's needed. It's the log evidence across all the samples, and it needs to be calculated at the end. For PG/SMC it's just the average log evidence, but IS has a slightly more complex method of calculating the log evidence. 

Because state variables are all stored in `spl.state`, it's the best way to get it to the `Chain` once the sampling is completed. Previously we just had some variable `logevidence` stored in each `sample` call which was not particularly modular.